### PR TITLE
Fix state bugs in active file analysis

### PR DESCRIPTION
**Pull request by sharwell**
Review commit-by-commit recommended

### CODE REVIEW

**Comment by genlu** [comment]
What bug does this change fix?

**Comment by sharwell** [comment]
We've had occasional reports of Current File analysis scope not behaving as expected, but without clear repro steps. While reviewing uses of this internally, I found cases where state was not correctly updated and caches were populated using information that needs to change when the analysis scope changes. This pull request fixes the cases I found.

File: src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs [hunk 1]
```diff
@@ -639,15 +676,17 @@ internal async Task Document_Change(BackgroundAnalysisScope analysisScope, bool
 
             var worker = await ExecuteOperation(workspace, w => w.ChangeDocument(document.Id, SourceText.From("//")));
 
+            var expectedDocumentEvents = 1;
```
Comment (mavasani) [inline src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs:679]: Why all these changes to tests? Do we now expect the same events in all the analysis modes?

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs [hunk 2]
```diff
@@ -320,17 +320,29 @@ public Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancel
...
 
+            if (!document.SupportsDiagnostics())
+            {
+                return false;
+            }
+
             // change it to check active file (or visible files), not open files if active file tracking is enabled.
             // otherwise, use open file.
-            return document.IsOpen() && document.SupportsDiagnostics();
+            if (SolutionCrawlerOptions.GetBackgroundAnalysisScope(document.Project) == BackgroundAnalysisScope.ActiveFile)
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs:338]: So active file analysis scope is no longer a solution crawler concept for all incremental analyzers, but just something that diagnostic incremental analyzer respects?

Comment (sharwell) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs:338]: Yes. The same already applied to `OpenFilesAndProjects` and `FullSolution`.

File: src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs:679]: Yes. Incremental analyzers are now responsible for choosing whether or not to act (and the scope of actions) for each analysis mode.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.IncrementalAnalyzerProcessor.cs [hunk 3]
```diff
@@ -77,7 +77,7 @@ private partial class IncrementalAnalyzerProcessor
                     }
 
                     // event and worker queues
-                    _documentTracker = _registration.Workspace.Services.GetService<IDocumentTrackingService>();
+                    _documentTracker = _registration.Workspace.Services.GetRequiredService<IDocumentTrackingService>();
```
Comment (mavasani) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.IncrementalAnalyzerProcessor.cs:80]: I thought this was possible to be null in the OOP process.

Comment (sharwell) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.IncrementalAnalyzerProcessor.cs:80]: It is no longer possible to be null. A default implementation is provided in which no document is ever active.

**Comment by mavasani** [comment]
Given the race conditions in the current version of implementation, I think it would be fine to take this approach.

**Review by mavasani — APPROVED** [review]

**Pull request merged**